### PR TITLE
Add `copy()` and `deepCopy()` methods to RobotTrajectory class.

### PR DESCRIPTION
**Pull request by shivaang12**
### Description

This PR points to issue #1660. I have deprecated the "normal" copy constructors and added two new methods, `RobotTrajectory::copy()` and `RobotTrajectory::deepCopy()`. Also documented these methods.

### Checklist
- [ ] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Document API changes relevant to the user in the MIGRATION.md notes
- [ ] Create tests, which fail without this PR [reference](https://ros-planning.github.io/moveit_tutorials/)
- [ ] Include a screenshot if changing a GUI
- [ ] While waiting for someone to review your request, please help review [another open pull request](https://github.com/ros-planning/moveit/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

File: moveit_core/robot_trajectory/src/robot_trajectory.cpp [hunk 1]
```diff
@@ -53,6 +53,26 @@ RobotTrajectory::RobotTrajectory(const robot_model::RobotModelConstPtr& robot_mo
...
+{
+  this->robot_model_ = robot_traj.robot_model_;
+  this->group_ = robot_traj.group_;
+  this->waypoints_ = robot_traj.waypoints_;
+  this->duration_from_previous_ = robot_traj.duration_from_previous_;
+}
+
+void RobotTrajectory::deepCopy(const RobotTrajectory& robot_traj)
+{
+  this->robot_model_ = robot_traj.robot_model_;
```
Comment (davetcoleman) [inline moveit_core/robot_trajectory/src/robot_trajectory.cpp]: Is this really a deep copy of the RobotModel?

File: moveit_core/robot_trajectory/src/robot_trajectory.cpp [hunk 2]
```diff
@@ -53,6 +53,26 @@ RobotTrajectory::RobotTrajectory(const robot_model::RobotModelConstPtr& robot_mo
...
+  this->group_ = robot_traj.group_;
+  this->waypoints_.resize(robot_traj.waypoints_.size());
+  for (auto it = robot_traj.waypoints_.begin(); it != robot_traj.waypoints_.end(); it++)
+  {
+    (this->waypoints_).emplace_back(std::make_shared<moveit::core::RobotState>(*(*it)));
```
Comment (davetcoleman) [inline moveit_core/robot_trajectory/src/robot_trajectory.cpp]:
```suggestion
    this->waypoints_.emplace_back(std::make_shared<moveit::core::RobotState>(*(*it)));
```
Are the () needed?

**Comment by shivaang12** [comment]
@davetcoleman I noticed there wasn't any unit test for the robot_trajectory package. Is it okay if I create it?

File: moveit_core/robot_trajectory/src/robot_trajectory.cpp (hunk 1 above)
Comment (shivaang12) [reply moveit_core/robot_trajectory/src/robot_trajectory.cpp]: Thanks for the review! No, this isn't a deep copy. But I observer that constructor is receiving a shared ptr so I thought I should assigned it to a same shared pointer. But now I get the picture, value should be same not the memory! Sorry, I am not much familiar with this part of the package.

Comment (v4hn) [reply moveit_core/robot_trajectory/src/robot_trajectory.cpp]: i believe we actually want to keep the same `RobotModel` in the copied trajectory.

the "deep" is supposed to imply that RobotStates are copied, not the whole underlying moveit class structure.

other proposals for this name are welcome.

File: moveit_core/robot_trajectory/include/moveit/robot_trajectory/robot_trajectory.h [hunk 3]
```diff
@@ -56,6 +56,18 @@ class RobotTrajectory
 
   RobotTrajectory(const robot_model::RobotModelConstPtr& robot_model, const robot_model::JointModelGroup* group);
 
+  RobotTrajectory(const RobotTrajectory&) = delete;
```
Comment (v4hn) [inline moveit_core/robot_trajectory/include/moveit/robot_trajectory/robot_trajectory.h]: please `[[deprecate]]` for now and forward it to `copy() `

**Comment by rhaschke** [comment]
Sorry for the late comment, but I think we could just implement the copy constructor as a deepcopy operation on the waypoints. Is there any use case for a shallow copy, i.e. keeping the waypoints but modifying the durations only? @v4hn: What do you think?

**Comment by v4hn** [comment]
This is the usual "should we change semantics without changing interfaces" question.
There's likely *someone* out there who exploited the shallow copy.

In this case I somewhat prefer the implemented approach (that I proposed),
but I don't have a strong opinion and it is indeed a rather drastic step
to eventually remove it as it breaks, e.g., STL containers...

Either way I would like to keep the proposed copy methods around.

**Comment by rhaschke** [comment]
> This is the usual "should we change semantics without changing interfaces" question.

I understand this argument in general, but it is rather unlikely in this case that somebody made use of this. In my point of view, it's a simple bug fix to return a deep copy of the waypoints by the copy constructor.

> Either way I would like to keep the proposed copy methods around.

IMHO, this just clutters the API in this case.

**Comment by v4hn** [comment]
> In my point of view, it's a simple bug fix to return a deep copy of the waypoints by the copy constructor.
> ...
> IMHO, this just clutters the API in this case.

We'll need a third opinion then, as I do not agree with you (nor do I strongly disagree though).

@henningkayser :-)

File: moveit_core/robot_trajectory/include/moveit/robot_trajectory/robot_trajectory.h (hunk 3 above)
Comment (shivaang12) [reply moveit_core/robot_trajectory/include/moveit/robot_trajectory/robot_trajectory.h]: cool!

Comment (shivaang12) [reply moveit_core/robot_trajectory/include/moveit/robot_trajectory/robot_trajectory.h]: I was thinking may be we can pass some message along with the deprecated attribute.

**Comment by davetcoleman** [comment]
> I noticed there wasn't any unit test dir for the robot_trajectory package. Is it okay if I create it?

Its always ok to add more tests! :-)

> We'll need a third opinion then, as I do not agree with you (nor do I strongly disagree though).

I lean towards @rhaschke 's thought that its a simple bug fix that makes the API more correct, and that should not be backported to Melodic but should be in MIGRATION.md

But I also want to hear @henningkayser 's opinion

**Comment by shivaang12** [comment]
I am curious to know if we have a decision here?

File: moveit_core/robot_trajectory/src/robot_trajectory.cpp (hunk 1 above)
Comment (henningkayser) [reply moveit_core/robot_trajectory/src/robot_trajectory.cpp]: > i believe we actually want to keep the same `RobotModel` in the copied trajectory.

I agree

> the "deep" is supposed to imply that RobotStates are copied, not the whole underlying moveit class structure.
> 
> other proposals for this name are welcome.

I think "deep" could be a bit misleading. I would prefer something more explicit, maybe `copyWithRobotStates()`.

**Comment by henningkayser** [comment]
> IMHO, this just clutters the API in this case.

Can't we just change the default copy constructor to shallow copy (here `copy()`) and just provide one alternative for specifically copying the robot states? I think it wouldn't clutter the API if the added functions don't appear redundant and a single added copy function doesn't seem too much to me.

**Comment by v4hn** [comment]
> Can't we just change the default copy constructor to shallow copy (here `copy()`) and just provide one alternative for specifically copying the robot states?

If I understand your comment right, it's the exact opposite of what Robert and Dave proposed. :-)
They argue `copy()` is not needed at all (I'm unsure about that) and `copyWithRobotStates` should be the semantics of the copy constructor.

**Comment by henningkayser** [comment]
> Can't we just change the default copy constructor to shallow copy (here `copy()`) and just provide one alternative for specifically copying the robot states?
> If I understand your comment right, it's the exact opposite of what Robert and Dave proposed. :-)
Well, you asked for my opinion ;)

> They argue `copy()` is not needed at all (I'm unsure about that)

I think shallow copy is probably the most needed of them all.
If I copy a trajectory I do this in order to modify the waypoint sequence, i.e. by adding or removing robot states. Deep copy is only needed if the actual waypoints are modified. Running different time parametrization algorithms (or similar processing methods) is the only valid use case I see here. In this case the semantics should be clear that you need a deep copy, but I wouldn't expect it as a default.

and `copyWithRobotStates` should be the semantics of the copy constructor.

**Comment by rhaschke** [comment]
> I think shallow copy is probably the most needed of them all.
If I copy a trajectory I do this in order to modify the waypoint sequence, i.e. by adding or removing robot states.

Good point. Adding or removing states doesn't require a deep copy.
So, we keep the shallow copy behavior in the (default) copy constructor and we add a new `deepCopy()` method to achieve deep copy. Naming it `copyWithRobotStates` is somewhat too verbose in my opinion.

**Comment by v4hn** [comment]
> Good point. Adding or removing states doesn't require a deep copy.
So, we keep the shallow copy behavior in the (default) copy constructor and we add a new deepCopy() method to achieve deep copy. Naming it copyWithRobotStates is somewhat too verbose in my opinion.

The problem with this solution is that users will probably still make the same silly mistake and perform a shallow copy, using the copy constructor - I did so two years ago and a colleague of mine did some months ago.
I believe nobody will read through the API first to discover the `deepCopy` method and infer that copy construction does something different.
This is the original pitfall I wanted to remove from the API with this patch...

**Comment by shivaang12** [comment]
I have revived the copy constructor which calls method copy(). Let me know if everyone is fine with it.

File: moveit_core/robot_trajectory/src/robot_trajectory.cpp (hunk 1 above)
Comment (mlautman) [reply moveit_core/robot_trajectory/src/robot_trajectory.cpp]: Perhaps `RobotTrajectory::copy(const RobotTrajectory& other, bool copy_robot_states = false)` would be more explicit. This is the approach used in `robot_state/conversions.h`

File: moveit_core/robot_trajectory/src/robot_trajectory.cpp [hunk 4]
```diff
@@ -53,6 +53,31 @@ RobotTrajectory::RobotTrajectory(const robot_model::RobotModelConstPtr& robot_mo
...
+  this->group_ = robot_traj.group_;
+  this->waypoints_ = robot_traj.waypoints_;
+  this->duration_from_previous_ = robot_traj.duration_from_previous_;
+}
+
+void RobotTrajectory::deepCopy(const RobotTrajectory& robot_traj)
+{
+  this->robot_model_ = robot_traj.robot_model_;
+  this->group_ = robot_traj.group_;
+  this->waypoints_.resize(robot_traj.waypoints_.size());
```
Comment (mlautman) [inline moveit_core/robot_trajectory/src/robot_trajectory.cpp]: This should not be used in combination with emplace_back

**Comment by rhaschke** [comment]
Additionally to @mlautman's fixups I restored the assignment operator (performing a shallow copy),
and implemented the copy operation(s) as a copy constructor with an additional deepcopy parameter that is false by default, thus maintaining the old behavior.

**Pull request merged**